### PR TITLE
Docker shell

### PR DESCRIPTION
**Pull request by ssalinas**
New PR against master for shell commands with docker

### CODE REVIEW

File: SingularityExecutor/src/main/java/com/hubspot/singularity/executor/shells/SingularityExecutorShellCommandDescriptor.java [hunk 1]
```diff
@@ -22,6 +24,10 @@
   @NotNull
   private List<SingularityExecutorShellCommandOptionDescriptor> options = Collections.emptyList();
 
+  @JsonProperty
+  @NotNull
+  private boolean switchUser = true;
```
Comment (tpetr) [inline SingularityExecutor/src/main/java/com/hubspot/singularity/executor/shells/SingularityExecutorShellCommandDescriptor.java]: can we make this more descriptive? (i.e. `runAsTaskUser`)

File: SingularityExecutor/src/main/java/com/hubspot/singularity/executor/shells/SingularityExecutorShellCommandRunner.java [hunk 2]
```diff
@@ -119,14 +124,39 @@ public boolean apply(SingularityExecutorShellCommandDescriptor input) {
 
     final SingularityExecutorShellCommandDescriptor shellCommandDescriptor = matchingShellCommandDescriptor.get();
 
-    List<String> command = new ArrayList<>(shellCommandDescriptor.getCommand());
+    List<String> command = new ArrayList<>();
+
+    if (shellCommandDescriptor.isSwitchUser()) {
+      String switchUserCommand = String.format(executorConfiguration.getSwitchUserCommandFormat(), taskProcess.getTask().getExecutorData().getUser().or(executorConfiguration.getDefaultRunAsUser()));
+      command.addAll(Arrays.asList(switchUserCommand.split("\\s+")));
```
Comment (tpetr) [inline SingularityExecutor/src/main/java/com/hubspot/singularity/executor/shells/SingularityExecutorShellCommandRunner.java]: seems a little janky, might be better to explicitly define what the switch user command for shell commands is

File: SingularityExecutor/src/main/java/com/hubspot/singularity/executor/shells/SingularityExecutorShellCommandRunner.java [hunk 3]
```diff
@@ -119,14 +124,39 @@ public boolean apply(SingularityExecutorShellCommandDescriptor input) {
...
 
     for (int i = 0; i < command.size(); i++) {
       if (command.get(i).equals(executorConfiguration.getShellCommandPidPlaceholder())) {
-        if (!taskProcess.getCurrentPid().isPresent()) {
-          throw new InvalidShellCommandException("No PID found");
+        int pid;
+        Path pidFilePath = MesosUtils.getTaskDirectoryPath(getTask().getTaskId()).resolve(executorConfiguration.getShellCommandPidFile());
+        if (Files.exists(pidFilePath)) {
+          try {
+            pid = Integer.parseInt(new Scanner(pidFilePath).useDelimiter("\\Z").next());
```
Comment (tpetr) [inline SingularityExecutor/src/main/java/com/hubspot/singularity/executor/shells/SingularityExecutorShellCommandRunner.java]: what is `\\Z` ?

File: SingularityExecutor/src/main/java/com/hubspot/singularity/executor/config/SingularityExecutorConfiguration.java [hunk 4]
```diff
@@ -247,6 +247,9 @@
   @JsonProperty
   private String shellCommandUserPlaceholder = "{USER}";
 
+  @JsonProperty
+  private String shellCommandPidFile = ".shell_command_pid";
```
Comment (tpetr) [inline SingularityExecutor/src/main/java/com/hubspot/singularity/executor/config/SingularityExecutorConfiguration.java]: this is the PID file for the command that the shell command should operate on, right? `.shell_command_pid` seems a little ambiguous to me, can we make it clearer? maybe `.task-pid`?

File: SingularityExecutor/src/main/java/com/hubspot/singularity/executor/shells/SingularityExecutorShellCommandRunner.java (hunk 2 above)
Comment (ssalinas) [inline SingularityExecutor/src/main/java/com/hubspot/singularity/executor/shells/SingularityExecutorShellCommandRunner.java]: Does it make sense to have another switch user command when we already configure the one that the executor uses for everything else? That's why I went with this originally.

File: SingularityExecutor/src/main/java/com/hubspot/singularity/executor/shells/SingularityExecutorShellCommandRunner.java (hunk 3 above)
Comment (ssalinas) [inline SingularityExecutor/src/main/java/com/hubspot/singularity/executor/shells/SingularityExecutorShellCommandRunner.java]: `\Z` is the end of string anchor, used that so we are essentially just read the whole file as one string

File: SingularityExecutor/src/main/java/com/hubspot/singularity/executor/config/SingularityExecutorConfiguration.java (hunk 4 above)
Comment (ssalinas) [inline SingularityExecutor/src/main/java/com/hubspot/singularity/executor/config/SingularityExecutorConfiguration.java]: :+1: updated

File: SingularityExecutor/src/main/java/com/hubspot/singularity/executor/shells/SingularityExecutorShellCommandRunner.java [hunk 5]
```diff
@@ -119,14 +123,36 @@ public boolean apply(SingularityExecutorShellCommandDescriptor input) {
...
+          } catch (Exception e) {
+            throw new InvalidShellCommandException("No PID found");
```
Comment (tpetr) [inline SingularityExecutor/src/main/java/com/hubspot/singularity/executor/shells/SingularityExecutorShellCommandRunner.java]: may want to clarify why the PID wasn't found by including the exception message

**Comment by tpetr** [comment]
:ship:

**Pull request merged**